autoPyTorch/data/utils.py

```diff
@@ -266,6 +368,84 @@ def reduce_precision(
     return X, reduced_dtypes, dtypes
 
 
+def subsample(
+    X: DatasetCompressionInputType,
+    is_classification: bool,
+    sample_size: Union[float, int],
+    y: Optional[SupportedTargetTypes] = None,
+    random_state: Optional[Union[int, np.random.RandomState]] = None,
+) -> Tuple[DatasetCompressionInputType, np.ndarray]:
+    """Subsamples data returning the same type as it recieved.
+
+    If `is_classification`, we split using a stratified shuffle split which
+    preserves unique labels in the training set.
+
+    NOTE:
+    It's highly unadvisable to use lists here. In order to preserve types,
+    we convert to a numpy array and then back to a list.
+
+    NOTE2:
+    Interestingly enough, StratifiedShuffleSplut and descendants don't support
+    sparse `y` in `split(): _check_array` call. Hence, neither do we.
+
+    Args:
+        X: DatasetCompressionInputType
+            The X's to subsample
+        y: SupportedTargetTypes
+            The Y's to subsample
+        is_classification: bool
+            Whether this is classification data or regression data. Required for
+            knowing how to split.
+        sample_size: float | int
+            If float, percentage of data to take otherwise if int, an absolute
+            count of samples to take.
+        random_state: int | RandomState = None
+            The random state to pass to the splitted
+
+    Returns:
+        (DatasetCompressionInputType, SupportedTargetTypes)
+            The X and y subsampled according to sample_size
+    """
+
+    if isinstance(X, List):
+        X = np.asarray(X)
+    if isinstance(y, List):
+        y = np.asarray(y)
+
+    if is_classification and y is not None:
+        splitter = CustomStratifiedShuffleSplit(
+            train_size=sample_size, random_state=random_state
+        )
+        left_idxs, _ = next(splitter.split(X=X, y=y))
+
+        if isinstance(X, pd.DataFrame):
+            idxs = X.index[left_idxs]
+            X = X.loc[idxs]
+        else:
+            X = X[left_idxs]
+
+        if isinstance(y, pd.DataFrame) or isinstance(y, pd.Series):
+            idxs = y.index[left_idxs]
+            y = y.loc[idxs]
+        else:
+            y = y[left_idxs]
```

Comment:
Could you make it as it makes this function easier to track.
```python
def _subsample_by_indices(
    X: Union[np.ndarray, pd.DataFrame],
    y: Union[np.ndarray, pd.DataFrame, pd.Series],
    indices_to_leave: np.ndarray
) -> Tuple[Union[np.ndarray, pd.DataFrame], Union[np.ndarray, pd.DataFrame, pd.Series]]:

    if ispandas(X):
        indices = X.index[indices_to_leave]
        X = X.loc[indices]
    else:
        X = X[indices_to_leave]

    if ispandas(y):
        indices = y.index[indices_to_leave]
        y = y.loc[indices]
    else:
        y = y[indices_to_leave]
```

Reply:
definitely.